File: Internship Tasks/Histogram Equalizer for WebCam/HEW.py
import cv2
import numpy as np


def pdf(image, histogram):
    for y in range(image.shape[0]):
        for x in range(image.shape[1]):
            intensity = image[y, x]
            histogram[intensity] += 1


def cdf(histogram, cumulative_histogram):
    cumulative_histogram[0] = histogram[0]

    for i in range(1, 256):
        cumulative_histogram[i] = histogram[i] + cumulative_histogram[i - 1]


def histogram_display(histogram, name):
    histogram_width = 1000
    histogram_height = 300
    bin_width = histogram_width // 256

    histogram_image = 255 * np.ones((histogram_height, histogram_width, 3), dtype=np.uint8)

    maximum_intensity = max(histogram)

    for i in range(256):
        height = int(histogram[i] / maximum_intensity * histogram_height)
        cv2.rectangle(histogram_image, (i * bin_width, histogram_height - height),
                      ((i + 1) * bin_width - 1, histogram_height - 1), (0, 0, 0), -1)

    cv2.imshow(name, histogram_image)
    cv2.waitKey(1)


camera = cv2.VideoCapture(0)

cv2.namedWindow("Original Image")
cv2.namedWindow("Equalized Image")
cv2.namedWindow("Original Histogram")
cv2.namedWindow("Equalized Histogram")

cv2.resizeWindow("Original Image", 640, 480)
cv2.resizeWindow("Equalized Image", 640, 480)
cv2.resizeWindow("Original Histogram", 1000, 300)
cv2.resizeWindow("Equalized Histogram", 1000, 300)

while True:
    ret, rgb_image = camera.read()

    image = cv2.cvtColor(rgb_image, cv2.COLOR_BGR2GRAY)

    histogram = [0] * 256
    size = image.shape[0] * image.shape[1]
    alpha = 255.0 / size

    pdf(image, histogram)

    PRk = [0.0] * 256
    for i in range(256):
        PRk[i] = histogram[i] / size

    cumulative_histogram = [0] * 256
    cdf(histogram, cumulative_histogram)

    Sk = [0] * 256
    for i in range(256):
        Sk[i] = int(cumulative_histogram[i] * alpha)

    PSk = [0.0] * 256
    for i in range(256):
        PSk[Sk[i]] += PRk[i]

    final_values = [0] * 256
    for i in range(256):
        final_values[i] = int(PSk[i] * 255)

    final_image = np.zeros((image.shape[0], image.shape[1]), dtype=np.uint8)
    for y in range(image.shape[0]):
        for x in range(image.shape[1]):
            final_image[y, x] = Sk[image[y, x]]

    cv2.imshow("Original Image", image)
    histogram_display(histogram, "Original Histogram")

    cv2.imshow("Equalized Image", final_image)
    histogram_display(final_values, "Equalized Histogram")

    cv2.imwrite("EqualizedImage.jpg", final_image)

    if cv2.waitKey(1) == 27:  # Press 'Esc' to exit
        break

camera.release()
cv2.destroyAllWindows()

File: Internship Tasks/Histogram Equalizer for WebCam/HEW.cpp
#include <iostream>
#include "opencv2/imgproc.hpp"
#include "opencv2/imgcodecs.hpp"
#include "opencv2/highgui.hpp"
#include <opencv2/opencv.hpp>
#include "opencv2/core.hpp"

using namespace std;
using namespace cv;

// Function to calculate Probability Distribution Function (PDF) of the input image
void PDF(Mat image, int histogram[]) {
    // Initialize all intensity values to 0
    for(int i = 0; i < 256; i++) {
        histogram[i] = 0;
    }

    // Calculate the number of pixels for each intensity value
    for(int y = 0; y < image.rows; y++) {
        for(int x = 0; x < image.cols; x++) {
            histogram[(int)image.at<uchar>(y,x)]++;
        }
    }
}

// Function to calculate Cumulative Distribution Function (CDF) of the input image
void CDF(int histogram[], int cumulativeHistogram[]) {
    cumulativeHistogram[0] = histogram[0];
    for(int i = 1; i < 256; i++) {
        cumulativeHistogram[i] = histogram[i] + cumulativeHistogram[i-1];
    }
}

// Function to display histogram
void histogramDisplay(int histogram[], const char* name) {
    int newHistogram[256];
    for(int i = 0; i < 256; i++) {
        newHistogram[i] = histogram[i];
    }
    // Histogram size
    int histogramWidth = 600;
    int histogramHeight = 300;
    // Creating "bins" for the range of 256 intensity values
    int binWidth = cvRound((double) histogramWidth/256);
    Mat histogramImage(histogramHeight, histogramWidth, CV_8UC1, Scalar(255, 255, 255));
    // Finding maximum intensity level in the histogram
    int maximumIntensity  = newHistogram[0];
    for(int i = 1; i < 256; i++){
        if(maximumIntensity < newHistogram[i]){
            maximumIntensity = newHistogram[i];
        }
    }
    // Normalizing histogram in terms of rows (y)
    for(int i = 0; i < 256; i++) {
        newHistogram[i] = ((double)newHistogram[i]/maximumIntensity)*histogramImage.rows;
    }
    // Drawing the intensity level line
    for(int i = 0; i < 256; i++) {
        line(histogramImage, Point(binWidth*(i), histogramHeight ), Point(binWidth*(i), histogramHeight - newHistogram[i]), Scalar(0,0,0), 1, 8, 0);
    }
    // Display
    namedWindow(name, WINDOW_AUTOSIZE);
    imshow(name, histogramImage);
}

int main()
{
    VideoCapture camera(0);
    Mat RGBimage;
    while(true){
        camera >> RGBimage;
        Mat image;
        cvtColor(RGBimage, image, COLOR_BGR2GRAY);
        imshow("Gray", image);
        char a;
        a == waitKey(33);
        if (a == 27){
            break;
        }
    }
	
    // Call function to create histogram
    int histogram[256];
    PDF(image, histogram);

    // Get the image size
    int size = image.rows * image.cols;
    float alpha = 255.0 / size;

    // Probability distribution for intensity levels
    float PRk[256];
    for(int i = 0; i < 256; i++) {
        PRk[i] = (double)histogram[i] / size;
    }

    // Call function to create cumulative histogram
    int cumulativeHistogram[256];
    CDF(histogram,cumulativeHistogram);

    // Scaling operation
    int Sk[256];
    for(int i = 0; i < 256; i++)
    {
        Sk[i] = cvRound((double)cumulativeHistogram[i] * alpha);
    }

    // Initializing equalized histogram
    float PSk[256];
    for(int i = 0; i < 256; i++) {
        PSk[i] = 0;
    }

    // Mapping operation
    for(int i = 0; i < 256; i++) {
        PSk[Sk[i]] += PRk[i];
    }

    // Rounding to get final values
    int finalValues[256];
    for(int i = 0; i < 256; i++) {
        finalValues[i] = cvRound(PSk[i]*255);
    }

    // Creating equalized image
    Mat finalImage = image.clone();
    
    for(int y = 0; y < image.rows; y++) {
        for(int x = 0; x < image.cols; x++) {
            finalImage.at<uchar>(y,x) = saturate_cast<uchar>(Sk[image.at<uchar>(y,x)]); 
        }
    }
    // Displaying source image
    namedWindow("Original Image");
    imshow("Original Image", image);

    // Displaying source image histogram
    histogramDisplay(histogram, "Original Histogram");

    // Displaying histogram equalized image
    namedWindow("Equalized Image");
    imshow("Equalized Image",finalImage);

    // Displaying equalized image histogram
    histogramDisplay(finalValues, "Equalized Histogram");

    // Saving image as a file
    imwrite("EqualizedImage.jpg", finalImage);
    waitKey();
    return 0;
}
